### DISCUSSION TITLE
What are workspaces?

### DISCUSSION THREAD

**Discussion by abusch** [wezterm/wezterm#2243] [Q&A] [answered]
The default LaunchMenu has a "Create new Workspace" item, but I can't find any explanation of what a workspace is in the documentation (or maybe I've missed it...). Is workspace the same as a local domain perhaps?

**Comment by MuhammedZakir** [comment, marked as answer]
A workspace is a collection of panes, tabs and windows. You can switch between workspaces, sort of like switching between windows. See #1531 for more details.

To understand it better, I recommend you to play around with it. You can find a simple config here [1]. Save that to a file and use that as the config file using `--config-file` cli option (`wezterm-gui --config-file /path/to/config start`). Then, do something -- create a few panes, tabs & windows and run some commands in some panes. Press `alt+9` and create a new workspace (this will also switch to the newly created workspace). Do something just like before. Switch between workspaces using `ctrl+n` or `ctrl+p`, or using `alt+9`. After this, you will understand what workspace is capable of. :-)

[1] https://wezfurlong.org/wezterm/config/lua/keyassignment/SwitchWorkspaceRelative.html

**Reply by abusch** [reply]
Thanks a lot for the answer and the link! Mapping it to tmux concepts definitely helped :) I'll have to just play with it.

**Reply by wez** [reply]
The example at the bottom of https://wezfurlong.org/wezterm/config/lua/gui-events/gui-startup.html may be helpful when it comes to automatically setting up a workspace

**Reply by Lenbok** [reply]
What is not clearly explained/documented is where does a workspace "live" in the situation where you use remote muxing. Does each workspace live on a single domain, or can you have a workspace with tabs/panes that span multiple domains (and if so, what does such a workspace look like when you connect to it remotely, especially if your local wezterm config doesn't know about the same domains that the remote muxer knows about). Is there even a concept of "connecting to a remote workspace"

**Reply by wez** [reply]
A window is associated with a workspace; that association is really just a copy of the workspace name.  The GUI will only show windows that belong to the active workspace.  You can think of the GUI as querying the multiplexer to determine which windows belong a workspace of a given name, and only showing those.

When using multiplexing, the GUI essentially unions the window/tab/pane information from each remote mux and makes it visible to the local mux running in the GUI.

Despite seeming like a "workspace contains windows" relationship, the actual model is "window is tagged with a workspace label string".

So if you connect to a remote mux its list of windows is unioned with the local list of windows and the set of current workspaces is produced by iterating that list and copying out the workspace labels from each window.  Then you can select any of those and the GUI will filter to the active workspace and only show the matches as mentioned above.

**Comment by owenstake** [comment]
Just like tmux session

**Comment by sxhmilyoyo** [comment]
Is there a way to show windows from multiple workspaces at the same time? After I select a new worksapce, the old one will hide.
